pkg/personname/examples_test.go

```diff
@@ -0,0 +1,126 @@
+package personname_test
+
+import (
+	"fmt"
+	"github.com/suyashkumar/dicom/pkg/personname"
+)
+
+// How to parse an existing PN value.
+func ExampleParse() {
+	// A Raw PN string value with Alphabetic, Ideographic, and Phonetic information.
+	rawPN := "Potter^Harry^James^^=哈利^波特^詹姆^^=hɛər.i^pɒ.tər^dʒeɪmz^^"
+
+	// To parse a PN value, use personName.Parse
+	parsedPN, err := personname.Parse(rawPN)
+	if err != nil {
+		panic(err)
+	}
+
+	// PN values are broken into three groups: Alphabetic, Phonetic, and Ideographic.
+	// which can be accessed through getter methods.
+	fmt.Println("ALPHABETIC:", parsedPN.Alphabetic.DCM())
+	fmt.Println("IDEOGRAPHIC:", parsedPN.Ideographic.DCM())
+	fmt.Println("PHONETIC:", parsedPN.Phonetic.DCM())
+
+	// Each Group info in turn has Getter methods for retrieving individual data.
+	fmt.Println("FAMILY NAME:", parsedPN.Alphabetic.FamilyName)
+	fmt.Println("GIVEN NAME: ", parsedPN.Alphabetic.GivenName)
+	fmt.Println("MIDDLE NAME:", parsedPN.Alphabetic.MiddleName)
+	fmt.Println("NAME PREFIX:", parsedPN.Alphabetic.NamePrefix)
+	fmt.Println("NAME SUFFIX:", parsedPN.Alphabetic.NameSuffix)
+
+	// To print the original Raw value, simply use the string method.
+	fmt.Println("ORIGINAL RAW:", parsedPN.DCM())
+
+	// Output (gotest comparison here is busted, maybe due to ideographs?):
+	// ALPHABETIC: Potter^Harry^James^^
+	// IDEOGRAPHIC: 哈利^波特^詹姆^^
+	// PHONETIC: hɛər.i^pɒ.tər^dʒeɪmz^^
+	// FAMILY NAME: Potter
+	// GIVEN NAME:  Harry
+	// MIDDLE NAME: James
+	// NAME PREFIX:
+	// NAME SUFFIX:
+	// ORIGINAL RAW: Potter^Harry^James^^=哈利^波特^詹姆^^=hɛər.i^pɒ.tər^dʒeɪmz^^
+}
+
+// How to create new PN value.
+func ExampleNew() {
+	// Create a new PN like so
+	pnVal := personname.Info{
+		Alphabetic: personname.GroupInfo{
+			FamilyName: "Potter",
+			GivenName:  "Harry",
+			MiddleName: "James",
+		},
+		// Add empty group that will not render its separators.
+		Ideographic: personname.GroupInfo{
+			NoNullSeparators: true,
+		},
+		// Add empty group that will not render its separators.
+		Phonetic: personname.GroupInfo{
+			NoNullSeparators: true,
+		},
+		// This will remove trailing '=' separators
+		NoNullSeparators: true,
+	}
+
+	// Print the string, should render as 'Potter^Harry^James^^'.
+	fmt.Println("PN 1:", pnVal.DCM())
```

Comment:
Yes, DICOM specifically allows for writing PN values and removing separators on the tail that have null values, which is part of why it's nice to have a set of helper functions for parsing them. Here is the relevant section of the DICOM spec:

> Any of the five components may be an empty string. The component delimiter shall be the caret "^" character (5EH). Delimiters are required for interior null components. Trailing null components and their delimiters may be omitted. Multiple entries are permitted in each component and are encoded as natural text strings, in the format preferred by the named person.

And for groups:

> For the purpose of writing names in ideographic characters and in phonetic characters, up to 3 groups of components (see Annexes H, I and J) may be used. The delimiter for component groups shall be the equals character "=" (3DH). The three component groups of components in their order of occurrence are: an alphabetic representation, an ideographic representation, and a phonetic representation.
> 
> Any component group may be absent, including the first component group. In this case, the person name may start with one or more "=" delimiters. Delimiters are required for interior null component groups. Trailing null component groups and their delimiters may be omitted.

The most common format I see in the wild is `Potter^Harry^^^` where only the Alphabetic group is present, has null separators, but the rest is elided. I'm sure the "default" convention varies wildly from facility-to-facility / software-to-software though.

In theory you can mix and match with different settings with different groups, so `=^^^^=` would be a legal PN value.

Reply:
Ah fantastic explanation with great standard citations, thank you! 

Since it is an option on the serialized representation to trim trailing empty values, I'm wondering from an API perspective if it makes sense to have two serialization methods (instead of specifying it when creating the object). I say this because it seems more like an option that the user may opt to choose when serializing the value, as opposed to an intrinsic property of the value itself. This also simplifies creation of the Info struct. So something like 

```go
DCM() string
DCMWithTruncatedEmptyValues() string
```

or swap them if we feel that the default should be to truncate and the non-default serialization should include them.

Just an API suggestion, I won't block the PR on it. Can always update it in a follow up PR to see how we like it.

Reply:
I try to think of these types with round-tripping in mind. The types are currently designed to support the following flow:

Decode PN -> Inspect Values -> Re-encode to the original value

If we remove the `NoNullSeparators` field in favor of deciding at serialization time, we lose two key abilities:

- Inspecting a decoded value to see whether it was using null separators.
- Agnostically inspecting the parsed values, then re-serializing them identically to the source in some other format. Users would have to carry around the raw stream, and if they were making alterations, but wanted to keep the format, come up with their own helper functions to analyze the original value to tell them which serialization format to use.

I like were this PR ended up, if you want to conform all your PN values to a certain format, you can use the following flow:

```
func ExampleInfo_WithNullSeparators() {
	rawVal := "Potter^Harry"

	parsedPN, err := personname.Parse(rawVal)
	if err != nil {
		panic(err)
	}

	fmt.Println("ORIGINAL   :", parsedPN.DCM())

	reformatted := parsedPN.WithNullSeparators()
	fmt.Println("REFORMATTED:", reformatted.DCM())

	// Output:
	// ORIGINAL   : Potter^Harry
	// REFORMATTED: Potter^Harry^^^=^^^^=^^^^
}
```

I think I'll take a look at making a more granular .WithFormat() option that allows you to granularly set each groups null separator value in a single method and push that to this PR, but I do strongly feel that those should be data object transformation methods rather than being serialization-time choices so that the user can feel confident that they can round-trip these values without having to be aware / concerned with formatting details unless they want to be.

Reply:
I see! A couple follow up questions, more for my understanding I think.

> If we remove the NoNullSeparators field in favor of deciding at serialization time, we lose two key abilities:
> 
> * Inspecting a decoded value to see whether it was using null separators.

Can you help shed a little more light on this use case? Typically, why would a caller want to know if null separators were used? Would it sufficient to have a helper function just for that? 

> * Agnostically inspecting the parsed values, then re-serializing them identically to the source in some other format. Users would have to carry around the raw string, and if they were making alterations, but wanted to keep the format, come up with their own helper functions to analyze the original value to tell them which serialization format to use.

Can you shed more light on why most callers would want this behavior? Is there some worry that writing the value back out without trailing null separators may cause a problem in the source system, even though it's conformant with the DICOM standard? In theory, I also see how it might be nice to not modify the value if it wasn't intrinsically changed when writing back out, but there are other places in the existing write library where we can't guarantee that anyway (and also, it seems to add some complexity to the implementation). I'm trying to understand more concretely what the benefits are of a perfect round trip (at the cost of implementation and API complexity).

Just trying to understand more of the use cases here. In my experience with these things, usually simpler implementations and simpler, opinionated APIs are usually worth some inflexibility because there's less surface for error/bugs and the behavior of the thing is easier to understand--so long as it's very clear what the thing does and how it behaves. That said, I'm less familiar with the downstream use cases of looking at and modifying person name in particular (de-identification pipelines come to mind first and foremost, but I think in those cases writing back out in a compliant but different format seems perfectly ok).